File: backend/apps/family/models/family.py
from django.db import models

from apps.shared.models import BaseModel


class FamilyProfile(BaseModel):
    user = models.OneToOneField(
        'users.User',
        on_delete=models.CASCADE,
        related_name='family',
    )
    family_name = models.CharField(max_length=100)
    city = models.CharField(max_length=100, help_text="Ob-havo API uchun shahar nomi")

    class Meta:
        db_table = 'family_profile'

    def __str__(self):
        return f"{self.family_name} ({self.user.email})"


class HealthCondition(BaseModel):
    """
    Reference table of health conditions (allergy, diabetes, etc).
    Populated via fixture.
    """

    class Category(models.TextChoices):
        ALLERGY = 'ALLERGY', 'Allergiya'
        DIABETES = 'DIABETES', 'Diabet'
        HEART = 'HEART', 'Yurak kasalligi'
        OBESITY = 'OBESITY', 'Semizlik'
        OTHER = 'OTHER', 'Boshqa'

    name = models.CharField(max_length=100)
    category = models.CharField(max_length=20, choices=Category.choices)
    description = models.TextField(blank=True)

    class Meta:
        db_table = 'family_health_condition'
        ordering = ['category', 'name']

    def __str__(self):
        return self.name


class FamilyMember(BaseModel):
    class Gender(models.TextChoices):
        MALE = 'MALE', 'Erkak'
        FEMALE = 'FEMALE', 'Ayol'

    family = models.ForeignKey(
        FamilyProfile, on_delete=models.CASCADE, related_name='members'
    )
    name = models.CharField(max_length=100)
    age = models.PositiveSmallIntegerField()
    gender = models.CharField(max_length=10, choices=Gender.choices)

    liked_recipes = models.ManyToManyField(
        'recipes.Recipe', related_name='liked_by', blank=True
    )
    disliked_recipes = models.ManyToManyField(
        'recipes.Recipe', related_name='disliked_by', blank=True
    )
    health_conditions = models.ManyToManyField(
        HealthCondition, related_name='members', blank=True
    )
    allergen_ingredients = models.ManyToManyField(
        'recipes.Ingredient',
        related_name='allergic_members',
        blank=True,
        help_text="Ingredientlarga allergiya (dukkakli, sut, tuxum va h.k.)",
    )

    class Meta:
        db_table = 'family_member'

    def __str__(self):
        return f"{self.name} ({self.family.family_name})"

File: mobile/lib/features/family/domain/family.dart
import 'package:equatable/equatable.dart';

class FamilyProfile extends Equatable {
  const FamilyProfile({
    required this.id,
    required this.familyName,
    required this.city,
    this.membersCount = 0,
  });

  final int id;
  final String familyName;
  final String city;
  final int membersCount;

  factory FamilyProfile.fromJson(Map<String, dynamic> json) => FamilyProfile(
        id: json['id'] as int,
        familyName: json['family_name'] as String,
        city: json['city'] as String,
        membersCount: (json['members_count'] as int?) ?? 0,
      );

  @override
  List<Object?> get props => [id, familyName, city, membersCount];
}

class HealthCondition extends Equatable {
  const HealthCondition({
    required this.id,
    required this.name,
    required this.category,
    this.description = '',
  });

  final int id;
  final String name;
  final String category; // ALLERGY, DIABETES, HEART, OBESITY, OTHER
  final String description;

  factory HealthCondition.fromJson(Map<String, dynamic> json) => HealthCondition(
        id: json['id'] as int,
        name: json['name'] as String,
        category: json['category'] as String,
        description: (json['description'] as String?) ?? '',
      );

  @override
  List<Object?> get props => [id, name, category];
}

class FamilyMember extends Equatable {
  const FamilyMember({
    required this.id,
    required this.name,
    required this.age,
    required this.gender,
    this.healthConditions = const [],
    this.allergenIngredientIds = const [],
    this.likedRecipeIds = const [],
    this.dislikedRecipeIds = const [],
  });

  final int id;
  final String name;
  final int age;
  final String gender; // MALE/FEMALE/OTHER
  final List<HealthCondition> healthConditions;
  final List<int> allergenIngredientIds;
  final List<int> likedRecipeIds;
  final List<int> dislikedRecipeIds;

  factory FamilyMember.fromJson(Map<String, dynamic> json) {
    final hcs = (json['health_conditions'] as List?)
            ?.map((e) => HealthCondition.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [];
    final allergens = (json['allergen_ingredients'] as List?)
            ?.map((e) => (e as Map)['id'] as int)
            .toList() ??
        [];
    return FamilyMember(
      id: json['id'] as int,
      name: json['name'] as String,
      age: json['age'] as int,
      gender: json['gender'] as String,
      healthConditions: hcs,
      allergenIngredientIds: allergens,
    );
  }

  @override
  List<Object?> get props => [id, name, age, gender, healthConditions];
}
